elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java

```diff
@@ -603,16 +604,120 @@ public RelationshipType getRelationshipType(Object entity, String relation) {
 
     /**
      * Get a type for a field on an entity.
+     * <p>
+     * If this method is called on a bean such as the following
+     * <pre>
+     * {@code
+     * public class Address {
+     *
+     *     {@literal @}Id
+     *     private Long id
+     *
+     *     private String street1;
+     *
+     *     private String street2;
+     * }
+     * }
+     * </pre>
+     * Then
+     * <pre>
+     * {@code
+     * getType(Address.class, "id") = Long.class
+     * getType(Address.class, "street1") = String.class
+     * getType(Address.class, "street2") = String.class
+     * }
+     * </pre>
+     * But if the ID field is not "id" and there is no such non-ID field called "id", i.e.
+     * <pre>
+     * {@code
+     * public class Address {
+     *
+     *     {@literal @}Id
+     *     private Long surrogateKey
+     *
+     *     private String street1;
+     *
+     *     private String street2;
+     * }
+     * }
+     * </pre>
+     * Then
+     * <pre>
+     * {@code
+     * getType(Address.class, "id") = Long.class
+     * getType(Address.class, "surrogateKey") = Long.class
+     * getType(Address.class, "street1") = String.class
+     * getType(Address.class, "street2") = String.class
+     * }
+     * </pre>
+     * If, however, the ID field is not "id" and there is a non-ID field called "id", i.e.
+     * <pre>
+     * {@code
+     * public class Address {
+     *
+     *     {@literal @}Id
+     *     private Long surrogateKey
+     *
+     *     private String id;
+     *
+     *     private String street;
+     * }
+     * }
+     * </pre>
+     * Then
+     * <pre>
+     * {@code
+     * getType(Address.class, "surrogateKey") = Long.class
+     * getType(Address.class, "id") = String.class
+     * getType(Address.class, "street") = String.class
+     * }
+     * </pre>
      *
      * @param entityClass Entity class
      * @param identifier  Field to lookup type
      * @return Type of entity
      */
     public Class<?> getType(Class<?> entityClass, String identifier) {
+        if (isIdType(entityClass, identifier)) {
+            return getIdType(entityClass);
+        }
+
         ConcurrentHashMap<String, Class<?>> fieldTypes = getEntityBinding(entityClass).fieldsToTypes;
         return fieldTypes == null ? null : fieldTypes.get(identifier);
     }
 
+    /**
+     * Returns whether or not a type-checked field is an ID of the bean.
+     * <p>
+     * The ID field of the bean is annotated by JAP {@code {@literal @}Id} annotation.
+     *
+     * @param entityClass Entity class
+     * @param identifier  Field to lookup type
+     *
+     * @return {@code true} if the field is an ID field
+     */
+    public boolean isIdType(Class<?> entityClass, String identifier) {
```

Comment:
If we remove the above scenario (ID as an attribute), I believe we can revert to your earlier commit where this function didn't exist and we instead used the existing isIdType function in the EntityBinding.

Reply:
Thanks. I will do that.

I believe that previous check `!ID_IDENTIFIER.equals(getEntityBinding(entityClass).getIdFieldName())` has a bug when id is not "id" but user is actually filtering on a non-ID field. The new helper method will fix that bug